File: Algorithm I/Binary Search - Day 1/P704.cpp
// Date: Sun, Sep 26, 2021
// 704. Binary Search (https://leetcode.com/problems/binary-search/)

// time complexity: O(logn)
int search(vector<int>& nums, int target) {
	int pivot, left = 0, right = nums.size() - 1;
	while (left <= right) {
		pivot = (left + right) / 2;
		if (nums[pivot] == target)
			return pivot;
		if (nums[pivot] < target)
			left = pivot + 1;
		else
			right = pivot - 1;
	}
	return -1;
}

File: LeetCoding-Challenge/March-2022/P704.java
package leetcodingchallenge.march2022;

/**
 * Date: March 26, 2022
 * 704. Binary Search
 * https://leetcode.com/problems/binary-search/
 */

public class P704 {
    // time: O(log(n))
    // space: O(1)
    public int search(int[] nums, int target) {
        int low = 0, high = nums.length - 1, mid;
        while (low <= high) {
            mid = (low + high) / 2;
            if (target == nums[mid]) return mid;
            if (target > nums[mid]) low = mid + 1;
            else high = mid - 1;
        }
        return -1;
    }
}
